### PR TITLE
Add queryExpression option to MongoDb Source

### PR DESCRIPTION
**Pull request by artembilan**
Starting with `SI-4.3.6` we can configure query expression in the MongoDb query DSL style.

* Upgrade to `SI-4.3.6`
* Add `MongodbSourceProperties#queryExpression` property, which has a priority over just `query` if configured
* Add test to demonstrate how MongoDb query DSL can be used via a new `queryExpression` property
* Mention a new property in the `README.adoc`
* Resolve deprecations in the tests according Spring Boot upgrade

### CODE REVIEW

**Comment by artembilan** [comment]
M-m-m, I think it would be better to review with `?w=-1`, since there were a lot of indents inconsistency.

File: spring-cloud-starter-stream-source-mongodb/src/test/java/org/springframework/cloud/stream/app/mongodb/source/MongodbSourceApplicationTests.java [hunk 1]
```diff
@@ -16,121 +16,176 @@
...
+							.poll(1, TimeUnit.SECONDS);
+			assertThat(received, nullValue());
 		}
+
 	}
 
+	@TestPropertySource(properties = {
+			"mongodb.query-expression=new BasicQuery('{ }')" +
+					".limit(1)" +
+					".with(new org.springframework.data.domain.Sort('greeting'))",
```
Comment (artembilan) [inline spring-cloud-starter-stream-source-mongodb/src/test/java/org/springframework/cloud/stream/app/mongodb/source/MongodbSourceApplicationTests.java:154]: @garyrussell , this `Sort` sample is for discussion.
I would like to include this `org.springframework.data.domain` package to SI by default here in the `MongoDbMessageSource`:
```
//Register MongoDB query API package so FQCN can be avoided in query-expression.
((StandardTypeLocator) typeLocator).registerImport("org.springframework.data.mongodb.core.query");
```
WDYT?

File: spring-cloud-starter-stream-source-mongodb/src/test/java/org/springframework/cloud/stream/app/mongodb/source/MongodbSourceApplicationTests.java [hunk 2]
```diff
@@ -16,121 +16,176 @@
...
+
 		@Test
 		public void test() throws InterruptedException {
-			Message<?> received = messageCollector.forChannel(source.output()).poll(2, TimeUnit.SECONDS);
-			assertThat(received, CoreMatchers.notNullValue());
-			assertThat(received.getPayload(), Matchers.instanceOf(String.class));
+			Message<?> received =
+					this.messageCollector
+							.forChannel(this.source.output())
+							.poll(2, TimeUnit.SECONDS);
```
Comment (garyrussell) [inline spring-cloud-starter-stream-source-mongodb/src/test/java/org/springframework/cloud/stream/app/mongodb/source/MongodbSourceApplicationTests.java]: We should probably increase these timers to 10 seconds.

File: spring-cloud-starter-stream-source-mongodb/src/test/java/org/springframework/cloud/stream/app/mongodb/source/MongodbSourceApplicationTests.java [hunk 3]
```diff
@@ -16,121 +16,176 @@
...
+
 		@Test
 		public void test() throws InterruptedException {
-			Message<?> received = messageCollector.forChannel(source.output()).poll(1, TimeUnit.SECONDS);
-			assertThat(received, CoreMatchers.nullValue());
+			Message<?> received =
+					this.messageCollector
+							.forChannel(this.source.output())
+							.poll(1, TimeUnit.SECONDS);
+			assertThat(received, nullValue());
```
Comment (garyrussell) [inline spring-cloud-starter-stream-source-mongodb/src/test/java/org/springframework/cloud/stream/app/mongodb/source/MongodbSourceApplicationTests.java]: This test seems to be bogus; it's the same as `DefaultTests` but with a long trigger - it seems to be just testing that the trigger doesn't fire within 1 second.

**Review by garyrussell — CHANGES_REQUESTED** [review]
That's all - yes I agree that we should add that package to the evaluation context within mongo only.

**Review by garyrussell — APPROVED** [review]

**Comment by garyrussell** [comment]
Merged as e7c50b556a6fa1364f4cec07058fba3c48063edd

**Pull request closed without merging**